src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientServices.cs

```diff
@@ -24,32 +25,217 @@ namespace Microsoft.Azure.IIoT.OpcUa.Protocol.Services {
+                TimeSpan.FromMinutes(2), TimeSpan.FromMinutes(2));
+            InitApplicationSecurityAsync().Wait();
+
+            _timer = new Timer(_ => OnTimer(), null, kEvictionCheck, Timeout.InfiniteTimeSpan);
+
+        }
+
+
+        /// <summary>
+        /// Create client host services
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <param name="configuration"></param>
+        /// <param name="maxOpTimeout"></param>
+        public ClientServices(ILogger logger,
+            IClientServicesConfig configuration,
+            TimeSpan? maxOpTimeout = null){
+
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _configuration = configuration ?? 
+                throw new ArgumentNullException(nameof(configuration)); ;
             _maxOpTimeout = maxOpTimeout;
-            _config = CreateApplicationConfiguration(TimeSpan.FromMinutes(2),
-                TimeSpan.FromMinutes(2));
-            Certificate = CertificateFactory.CreateCertificate(
-                _config.SecurityConfiguration.ApplicationCertificate.StoreType,
-                _config.SecurityConfiguration.ApplicationCertificate.StorePath, null,
-                _config.ApplicationUri, _config.ApplicationName,
-                _config.SecurityConfiguration.ApplicationCertificate.SubjectName, null,
-                CertificateFactory.defaultKeySize, DateTime.UtcNow - TimeSpan.FromDays(1),
-                CertificateFactory.defaultLifeTime, CertificateFactory.defaultHashSize,
-                false, null, null);
-            _timer = new Timer(_ => OnTimer(), null, kEvictionCheck,
-                Timeout.InfiniteTimeSpan);
+
+            // Create discovery config and client certificate
+            _OpcApplicationConfig = CreateApplicationConfiguration(
+                TimeSpan.FromMinutes(2), TimeSpan.FromMinutes(2));
+
+            InitApplicationSecurityAsync().Wait();
+
+            _timer = new Timer(_ => OnTimer(), null, kEvictionCheck, Timeout.InfiniteTimeSpan);
+        }
+
+
+        /// <summary>
+        /// Initialize the OPC UA Application's security configuration
+        /// </summary>
+        /// <returns></returns>
+        private async Task InitApplicationSecurityAsync(){
+
+
+            _OpcApplicationConfig.SecurityConfiguration.AddAppCertToTrustedStore = true;
+
+            X509Certificate2 ownCertificate = await _OpcApplicationConfig.SecurityConfiguration.
+                ApplicationCertificate.Find(true).ConfigureAwait(false);
+
+            if (ownCertificate == null) {
+
+                _logger.Information($"No existing Application own certificate found." +
+                    $" Create a self-signed Application certificate valid from yesterday " +
+                    $"for {CertificateFactory.defaultLifeTime} months,");
+                _logger.Information($"with a {CertificateFactory.defaultKeySize} bit " +
```

Comment:
Structured logging comment above. - and all other log output

Reply:
done